text/0000-light-dom.md

```diff
@@ -0,0 +1,192 @@
+  }
+}
+```
+
+In LWC as well, the developer will retain the choice: they can choose to inherit from `LightningElement` to select Shadow DOM (which already exists) or inherit from a new class, `MacroElement`, to opt into Light DOM.
+
+```js
+import { MacroElement } from "lwc";
+
+export default class MyLightComponent extends MacroElement {
+  renderedCallback() {
+    this.querySelector("child-element"); // note the absence of this.template
+  }
+}
+```
+
+`MacroElement` is a new class that will be exported from the `lwc` module. Its API and functionality remain identical to `LightningElement` with the primary difference of not using Shadow DOM.
+
+### Component features when using Light DOM
+
+Some of the LWC component capabilities are directly inherited from Shadow DOM, or emulated by the synthetic-shadow. Despite the use of Light DOM, we’d like to keep these features available, even if their behavior is slightly adapted to the Light DOM:
+
+- **Slots**
+
+  As we mentioned before, the component composition model in LWC is provided by slots. Light DOM will provide the same mental model for developers building Light DOM components.
+
+  In Light DOM, `<slot>` will denote the place where the slotted component will be attached. The `<slot>` element itself won't be rendered. The slotted content (or the fallback content) will be flattened to the parent element at runtime.
+
+  Since the `<slot>` element itself isn't rendered, adding attributes or event listeners to the `<slot>` element in the template will throw a compiler error.
+
+- **Scoped Styles**
+
+  Shadow DOM styles are scoped to the enclosing shadow tree. Styles don't leak out to the rest of the page and the page styles don't leak into this component. In Native shadow, it is enforced by the browser whereas in Synthetic shadow it is enforced by adding few attributes to the elements and scoping all CSS rules by those attributes.
+
+  Styles scoping in Light DOM will be done differently. The styles will be scoped not just to the component they belong, but also to its children. These styles won't leak out to the parents (or their children) though.
+
+  E.g.
+
+  ```html
+  <x-a>
+    <style>
+      p {
+        color: blue;
+      }
+    </style>
+
+    <p></p> <!-- will be blue -->
+    <p class="red"></p> <!-- will still be blue. red class from child won't leak out here -->
+
+    <x-b>
+      <style>
+        p.red {
+          color: red;
+        }
+      </style>
+
+      <p></p> <!-- will be blue. Style from x-a will be applied -->
+      <p class="red"></p> <!-- will be red -->
+      <p class="red"></p> <!-- This paragraph is coming from A and slotted into B. Will also be red. -->
+    </x-b>
+  </x-a>
+  ```
+
+  Opting out of this scoping is not supported. There's no way for a component author to say a CSS rule shouldn't be scoped to that specific component (and its children). If global scoping is desired, a global stylesheet can be injected manually.
```

Comment:
With synthetic shadow DOM, we implemented a system for style scoping similar to this (using attribute selectors). However, with shadow DOM, it was kind of implied that eventually we would migrate to native shadow, so the synthetic system wasn't designed to last forever. With this system, though, we are committing to support a new styling system indefinitely, including whatever implementation details we land on, which will impact compatibility moving forward. For instance:

1. Whether to use attribute selectors / classes / etc. for scoping? This has performance implications (browsers tend to have more optimizations for class than for attribute selectors) as well as specificity implications (e.g. specificity conflicts with global styles). FWIW, Vue uses attribute selectors whereas Svelte uses class selectors.
2. Whether styles are inherited to the children or not. This also seems to me to have performance implications, since I'm not sure how we could implement this, except by concatenating attributes down the descendant tree. (So for instance, would the 5th light DOM grandchild have 5 separate attributes on every DOM node?)
3. Whether styles are inherited to light DOM "slots" or not, and whether something like (for instance) sibling selectors would work across both "slot" and non-"slot" light DOM children

It seems to me that the safest option is to not provide any kind of style scoping for light DOM elements. I.e. "if you want style scoping, use shadow DOM." This avoids making any indefinite commitments toward one particular styling system in light DOM.

Reply:
I think the goal is to land on the most natural ergonomics for the developer, including developers already versed in LWC. 

Styles in LWC are scoped today, but for end users this is not a natural consequence of the shadow DOM encapsulation, rather it just makes sense since the authoring model is per-component. This is why Vue scopes styles, not because it aligns with the Shadow DOM model. For us that is just a bonus.

Not scoping styles would mean that without any extra code on our side, styles would leak into shadow DOM children, which would not match the native shadow DOM behavior when we turn off synthetic.

Reply:
@tedconn 

> Styles in LWC are scoped today, but for end users this is not a natural consequence of the shadow DOM encapsulation

It is due to `synthetic-shadow`, sure, but the long-term goal is to migrate to native shadow DOM. So then it would be a natural consequence. 🙂 

> This is why Vue scopes styles, not because it aligns with the Shadow DOM model.

I'm not saying scoped styles aren't good for developer ergonomics; I cede that point. 🙂  My point is just that "which scoped styles?" is an important question, especially since we have to indefinitely maintain whatever we choose today.

> Not scoping styles would mean that without any extra code on our side, styles would leak into shadow DOM children

I'm not sure I follow this. Today, in a purely non-LWC, non-synthetic world, I can have a light DOM-using component with a shadow DOM-using component as its child, and the only styles that leak in are [inheritable styles](https://open-wc.org/guides/knowledge/styling/styles-piercing-shadow-dom/) like `font-family` and `color`, which is per the spec. It should work this way both in native shadow DOM and in our synthetic shadow DOM.

Reply:
> It should work this way both in native shadow DOM and in our synthetic shadow DOM.

I may be wrong but my thinking was that If styles aren't scoped, in synthetic shadow DOM, those styles will leak in, because there's no other mechanism to prevent that.

Reply:
I wasn't sure, so I just ran a quick check. It looks like if you add the style `body { font-family: monospace; }` to a Salesforce page, it will indeed leak in to synthetic shadow DOM components. (This is the same as what would happen with a light DOM-using component wrapping a shadow DOM-using component.) So it appears we are following the spec correctly on that.

Reply:
>  those styles will leak in, because there's no other mechanism to prevent that.

@tedconn You can prevent inherited styles to be applied by resetting all the properties on the host element. It works in both native shadow and synthetic shadow. 

```css
:host {
  all: initial;
}
```

----

We discussed CSS scoping at length and we reached the conclusion that it something we would like to preserve with Light DOM for developer ergonomic reasons, as @tedconn mentioned. 

We considered the following approach to style scoping:

1. **component content scoping:** Styles are applied to the content the component renders and aren't applied to children components. This approach is similar to the way styles are scoped today in synthetic shadow. It also behaves the same as Vue, Svelte, and Angular. An HTML attribute is added to all the CSS selectors, the same attribute is also added to all the elements rendered by the component.

2. **component subtree scoping:** Styles are applied to the content the component renders and also to the content rendered in the children's components. This is a similar approach to Aura styling.  All the CSS selectors are prefixed with a CSS attribute selector and the same attribute is applied to the root element to scope the subtree.

```css
/* Input */
.foo .bar {}

/* Ouput: component content scoping */
.foo[data-scope] .bar[data-scope] {}

/* Output: component subtree scoping */
[data-scope] .foo .bar {}
```

We decided to go with the **component subtree scoping** approach for the following reasons:
- This approach map with the way `querySelector` works in the light DOM. It applies to the entire subtree and not only the current element.
- One of the main reason for using light DOM is to let a component author override it children components style. This is possible with the subtree scoping approach but not with the content scoping one. Overriding children component styles with the subtree scoping approach would require the introduction of a non-standard CSS selector to escape the scoping (`>>>` in [Vue](https://vue-loader.vuejs.org/guide/scoped-css.html#deep-selectors) and [Angular](https://angular.io/guide/component-styles#deprecated-deep--and-ng-deep) or `:global` in [Svelte](https://svelte.dev/docs#style)).

The main downside for me of the  **component subtree scoping** approach is that you can run into a style clashing issue where the parent component can unintentionally override children's component styles because the parent selector has a higher specificity.

Reply:
Thanks, that makes sense. Two comments:

First, I'm not sure that preferring Aura's styling model to Vue/Svelte's styling model is the best choice from the perspective of developer ergonomics. Aura is well-known inside Salesforce, but not so much outside. (Also if I'm not mistaken, most React CSS-in-JS libraries follow something more like the Vue/Svelte model.) Personally I would find it a bit surprising that parent light components can affect child light components' styles.

Second, for this CSS:

```css
[data-scope] .foo .bar {}
```

I find this to be a bit scary from a performance perspective. Imagine something like:

```css
/* Input */
div {}

/* Output */
[data-scope] div {}
```

Now you're potentially the telling the browser to check the ancestor chain for every `div` to see if its attributes match. For class selectors, there is already a Bloom filter to handle this, but I'm not sure if browsers have gotten around to optimizing attribute selectors in the same way. I may have to whip up a benchmark to test this. 🙂

Reply:
> Personally I would find it a bit surprising that parent light components can affect child light components' styles.

This is good to know. We can survey developers internally to see which approach they find the most intuitive.

> Now you're potentially the telling the browser to check the ancestor chain for every div to see if its attributes match. 

I don't think we can use classes because component developers can set classes via the template (from outside) or in the component via `this.classList` (from within). We would need to special-case this scoping class and ensure that it is always present even when the author manipulates the component classes. I am not sure if we can do this in an efficient way.